Code/Tools/AssetProcessor/native/AssetManager/assetScannerWorker.cpp

```diff
@@ -74,64 +79,118 @@ void AssetScannerWorker::ScanForSourceFiles(const ScanFolderInfo& scanFolderInfo
-        {
-            return;
-        }
+    QString intermediateAssetsFolder = QString::fromUtf8(AssetUtilities::GetIntermediateAssetsFolder(cachePath).c_str());
+    QString normalizedIntermediateAssetsFolder = AssetUtilities::NormalizeDirectoryPath(intermediateAssetsFolder);
 
-        QString absPath = entry.absoluteFilePath();
-        const bool isDirectory = entry.isDir();
-        QDateTime modTime = entry.lastModified();
-        AZ::u64 fileSize = isDirectory ? 0 : entry.size();
-        AssetFileInfo assetFileInfo(absPath, modTime, fileSize, &rootScanFolder, isDirectory);
+    // Implemented non-recursively so that the above functions only have to be called once per scan,
+    // and so that the performance is easy to analyze in a profiler:
+    QList<ScanFolderInfo> pathsToScanQueue;
+    pathsToScanQueue.push_back(scanFolderInfo);
 
-        // Filtering out excluded files
-        if (m_platformConfiguration->IsFileExcluded(absPath))
+    while (!pathsToScanQueue.empty())
+    {
+        ScanFolderInfo pathToScan = pathsToScanQueue.back();
+        pathsToScanQueue.pop_back();
+        QDir dir(pathToScan.ScanPath());
+        dir.setSorting(QDir::Unsorted);
+        // Only scan sub folders if recurseSubFolders flag is set
+        if (!scanFolderInfo.RecurseSubFolders())
         {
-            m_excludedList.insert(AZStd::move(assetFileInfo));
-            continue;
+            entries = dir.entryInfoList(QDir::NoDotAndDotDot | QDir::Files);
         }
-
-        if (isDirectory)
+        else
         {
-            //Entry is a directory
-            // The AP needs to know about all directories so it knows when a delete occurs if the path refers to a folder or a file
-            m_folderList.insert(AZStd::move(assetFileInfo));
-            ScanFolderInfo tempScanFolderInfo(absPath, "", "", false, true);
-            ScanForSourceFiles(tempScanFolderInfo, rootScanFolder);
+            entries = dir.entryInfoList(QDir::Dirs | QDir::NoDotAndDotDot | QDir::Files);
         }
-        else if (!AssetUtilities::IsInCacheFolder(absPath.toUtf8().constData())) // Ignore files in the cache
+
+        for (const QFileInfo& entry : entries)
         {
-            //Entry is a file
-            m_fileList.insert(AZStd::move(assetFileInfo));
+            if (!m_doScan) // scan was cancelled!
+            {
+                return;
+            }
+
+            QString absPath = entry.absoluteFilePath();
+            const bool isDirectory = entry.isDir();
+            QDateTime modTime = entry.lastModified();
+            AZ::u64 fileSize = isDirectory ? 0 : entry.size();
+            AssetFileInfo assetFileInfo(absPath, modTime, fileSize, &rootScanFolder, isDirectory);
+            QString relPath = absPath.mid(rootScanFolder.ScanPath().length() + 1);
+
+            if (isDirectory)
+            {
+                // in debug, assert that the paths coming from qt directory info iteration is already normalized
```

Comment:
The problem is that  these checks actually cost a lot of time and its the hot path of the application.
AssetUtilities::NormalizeDirectoryPath(absPath) is not free and is in fact what we're trying to avoid having to do.